### PR TITLE
updates to UN SDG scripts

### PR DESCRIPTION
**Pull request by n-h-diaz**
checking in all the latest scripts

### CODE REVIEW

File: scripts/un/sdg/geography.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+'''
+CONTAINMENT_TEMPLATE = '''
+Node: dcid:{dcid}
+typeOf: dcid:{type}{containment}
+'''
+
+# Curated map of SDG GEOGRAPHY_CODE to UN data code.
+FIXED = {
+    'africa': '2',
+    'undata-geo/G99999999': '952',
```
Comment (pradh) [inline scripts/un/sdg/geography.py:46]: nit:  comment what dcid this is

File: scripts/un/sdg/geography.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    Args:
+        file: Input file path.
+
+    Returns: 
+        - Map of UN code -> SDG code.
+        - Map of SDG code -> UN code.
+    '''
+    un2sdg = {}  # Map of UN code -> SDG code.
+    sdg2un = {}  # Map of SDG code -> UN code.
+    with open(file, encoding='utf-8-sig') as f:
```
Comment (pradh) [inline scripts/un/sdg/geography.py:79]: comment on choice of encoding?

File: scripts/un/sdg/geography.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    '''Produces map of UN code -> curated (dcid, DC type, DC name).
+
+    Args:
+        file: Input file path.
+
+    Returns:
+        Map of UN code -> curated (dcid, DC type, DC name).
+    '''
+    un2dc = {}
+    with open(file) as f:
```
Comment (pradh) [inline scripts/un/sdg/geography.py:101]: Is there some place that all the input structures are documented that we can link to?

File: scripts/un/sdg/geography.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    '''
+    un2dc2 = {}
+    new_subjects = set()
+    with open(input_geos) as f_in:
+        with open(output, 'w') as f_out:
+            reader = csv.DictReader(f_in)
+            for row in reader:
+                subject = row['subject_id']
+                if subject in un2dc:
+                    dcid = un2dc[subject][0]
```
Comment (pradh) [inline scripts/un/sdg/geography.py:173]: nit: dcid, type, name = un2dc[subject] ?

File: scripts/un/sdg/geography.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+                    dcid = row['subject_id'].replace(':', '/')
+                    if row['subject_id'] in un2sdg and un2sdg[
+                            row['subject_id']] in sdg2type:
+                        sdg_type = sdg2type[un2sdg[row['subject_id']]]
+                        if sdg_type == 'SamplingStation' or sdg_type == 'City':
+                            type = sdg_type
+                        else:
+                            type = 'GeoRegion'
+                    else:
+                        type = 'GeoRegion'
```
Comment (pradh) [inline scripts/un/sdg/geography.py:186]: use constants?

File: scripts/un/sdg/geography.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        containment: Map of (subject dcid, subject type) -> list of containing
+            object dcids.
+        new_subjects: Set of (dcid, type) for new places.
+
+    '''
+    with open(output, 'w') as f:
+        for s in sorted(containment):
+            c = ''
+            for o in containment[s]:
+                c += '\ncontainedInPlace: dcs:' + o
```
Comment (pradh) [inline scripts/un/sdg/geography.py:258]: nit: dcid:

File: scripts/un/sdg/geography.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+            c = '\ncontainedInPlace: dcs:Earth'
+            f.write(
+                CONTAINMENT_TEMPLATE.format_map({
+                    'dcid': s[0],
+                    'type': s[1],
+                    'containment': c
+                }))
+
+
+def write_place_mappings(output, sdg2un, un2dc, un2dc2):
```
Comment (pradh) [inline scripts/un/sdg/geography.py:280]: nit: use more meaningful names?

File: scripts/un/sdg/geography.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+            object = row['object_id'].removeprefix('sdg-geo:')  # SDG code.
+            if not subject or not object:
+                continue
+            un2sdg[subject] = object
+            sdg2un[object] = subject
+    return un2sdg, sdg2un
+
+
+def get_un2dc(file):
+    '''Produces map of UN code -> curated (dcid, DC type, DC name).
```
Comment (pradh) [inline scripts/un/sdg/geography.py:92]: for future, encourage using typing and dataclass (Node with 3 fields)

File: scripts/un/sdg/util.py [hunk 9]
```diff
@@ -211,6 +233,43 @@ def is_valid(v):
...
+    '''Curates PVs based on custom mappings.
+
+    Args:
+      text: Input text.
+      mappings: Custom mappings.
+
+    Returns: 
+      Formatted text.
+    '''
+    pairs = text[1:-1].split('|')
```
Comment (pradh) [inline scripts/un/sdg/util.py:246]: Some examples might help.

File: scripts/un/sdg/util.py [hunk 10]
```diff
@@ -95,57 +91,80 @@
...
-  Args:
-    file: Path to input file.
-
-  Returns:
-    City map.
-  '''
-    with open(file) as f:
-        reader = csv.DictReader(f)
-        return {row['name']: row['dcid'] for row in reader}
+# Series that should be dropped.
```
Comment (pradh) [inline scripts/un/sdg/util.py:114]: Add context for why?  A TODO to follow up...

File: scripts/un/sdg/util.py [hunk 11]
```diff
@@ -95,57 +91,80 @@
...
-REGIONS = {
-    1: 'Earth',
-    2: 'africa',
-    5: 'southamerica',
-    9: 'oceania',
-    10: 'antarctica',
-    21: 'northamerica',
-    142: 'asia',
-    150: 'europe',
+# Series where zero should be treated as null and removed.
```
Comment (pradh) [inline scripts/un/sdg/util.py:94]: nit: add context

File: scripts/un/sdg/process.py [hunk 12]
```diff
@@ -255,6 +324,18 @@ def process(input_dir, schema_dir, csv_dir):
 
                     val = 'SDG_' + enum + 'Enum_' + val
                     cprops += f'\n{prop}: dcs:{val}'
+                sources = df.loc[df['VARIABLE_CODE'] == row['VARIABLE_CODE']]
+                sources = sources.loc[:, ['SOURCE']].drop_duplicates()['SOURCE']
+                footnote = ''
+                if not sources.empty:
+                    footnote = '\nfootnote: "Includes data from the following sources: ' + '; '.join(
```
Comment (pradh) [inline scripts/un/sdg/process.py:331]: add a comment?

File: scripts/un/sdg/process.py [hunk 13]
```diff
@@ -96,7 +93,60 @@ def get_measurement_method(row):
...
+def fix_encoding(s):
+    '''Fixes input encoding to decode special characters.
+
+    Args:
+        s: Input string.
+
+    Returns:
+        Formatted string.
+    '''
+    try:
```
Comment (pradh) [inline scripts/un/sdg/process.py:143]: comment?

File: scripts/un/sdg/process.py [hunk 14]
```diff
@@ -27,54 +27,51 @@
...
 
     Returns:
-        Unit dcid.
+        Geography dcid.
     '''
-    if util.is_valid(base_period):
-        return f'[{units} {base_period}]'
-    return 'dcs:SDG_' + units
+    if str(code) in place_mappings:
+        return 'dcs:' + place_mappings[str(code)]
```
Comment (pradh) [inline scripts/un/sdg/process.py:73]: nit: place references should be `dcid:`?

**Review by pradh — APPROVED** [review]
Looks good!  A bunch of nits...

File: scripts/un/sdg/README.md [hunk 15]
```diff
@@ -23,7 +29,9 @@ Produces:
   * unit.mcf
 * csv/ folder: 
   * [CODE].csv
-(Note that the `schema/` folder is not included in the repository but can be regenerated by running the script.)
+(Note that these folders are not included in the repository but can be regenerated by running the script.)
+
+When refreshing the data, the `geography`, `schema`, and `csv` folders might all get updated and will need to be resubmitted to g3. The corresponding TMCF file is `sdg.tmcf`.
 
 To run unit tests:
```
Comment (pradh) [inline scripts/un/sdg/README.md:36]: Please consider adding caveats / gotchas to the README for context / follow-ups (like the fact that we drop some series)

File: scripts/un/sdg/geography.py (hunk 1 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py:48]: done

File: scripts/un/sdg/geography.py (hunk 2 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py:121]: done

File: scripts/un/sdg/geography.py (hunk 3 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py:143]: updated the structures and added comment

File: scripts/un/sdg/geography.py (hunk 5 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py]: done

File: scripts/un/sdg/util.py (hunk 10 above)
Comment (n-h-diaz) [reply scripts/un/sdg/util.py]: done

File: scripts/un/sdg/util.py (hunk 11 above)
Comment (n-h-diaz) [reply scripts/un/sdg/util.py]: done

File: scripts/un/sdg/process.py (hunk 13 above)
Comment (n-h-diaz) [reply scripts/un/sdg/process.py:143]: done

File: scripts/un/sdg/README.md (hunk 15 above)
Comment (n-h-diaz) [reply scripts/un/sdg/README.md:36]: done

File: scripts/un/sdg/process.py (hunk 14 above)
Comment (n-h-diaz) [reply scripts/un/sdg/process.py]: done

File: scripts/un/sdg/process.py (hunk 12 above)
Comment (n-h-diaz) [reply scripts/un/sdg/process.py:334]: done

File: scripts/un/sdg/geography.py (hunk 6 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py]: done

File: scripts/un/sdg/geography.py (hunk 7 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py]: done

File: scripts/un/sdg/geography.py (hunk 8 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py]: done

File: scripts/un/sdg/geography.py (hunk 4 above)
Comment (n-h-diaz) [reply scripts/un/sdg/geography.py]: switched to use class

File: scripts/un/sdg/util.py (hunk 9 above)
Comment (n-h-diaz) [reply scripts/un/sdg/util.py:251]: done

**Pull request merged**